master/internal/db/postgres_tasks.go

```diff
@@ -98,11 +98,22 @@ VALUES (:task_id, :allocation_id, :slots, :resource_pool, :agent_label, :start_t
 
 // CompleteAllocation persists the end of an allocation lifetime.
 func (db *PgDB) CompleteAllocation(a *model.Allocation) error {
-	return db.namedExecOne(`
+	if _, err := db.sql.Exec(`
+UPDATE allocations
+SET start_time = cluster_heartbeat FROM cluster_id
+WHERE start_time = $1`, time.Time{}); err != nil {
+		return errors.Wrap(err,
+			"setting start time to cluster heartbeat when it's assigned to zero value")
+	}
+
+	err := db.namedExecOne(`
 UPDATE allocations
 SET end_time = :end_time
 WHERE allocation_id = :allocation_id
 `, a)
+
+	return err
+
```

Comment:
i'm not sure i understand these changes, and the others in this file. i think we should just be setting start time in allocation.go and saving it.

Reply:
These changes are related to these concerns we had when we initially discussed the implementation: - CompleteAllocation if start time is zero value set to cluster_heartbeat. This can happen when we call cleanup after a Poststop message. This message is sent after panic or error. 
- CloseOpenAllocations if start time is zero value set to cluster_heartbeat.  This can happen when cluster fails. 

If we don't do this then we can end up with a zero value in start time and end_time as the current value or cluster_heartbeat which is incorrect.